query/grouped_result.go

```diff
@@ -0,0 +1,99 @@
+package query
+
+import "github.com/xichen2020/eventdb/document/field"
+
+// TransformFn transforms a list of raw calculated values into a final calculated value.
+// For example, computation of `Avg` should be transformed into computation of `Sum` and
+// `Count`, and once the results are gathered, the final `Avg` value can be computed as
+// `Sum` / `Count`.
+type TransformFn func(rawValues []float64) float64
+
+// CalculatedValue represents a calculated value. In simple cases this is simply a single
+// numeric value. However, there are also cases where calculation of a value requires
+// calculations of other intermediate values, which eventually get transformed into the
+// final calculated value (e.g., the average value of a field).
+type CalculatedValue struct {
+	Value       float64
+	Values      []float64
+	TransformFn TransformFn
+}
+
+// ResultGroup is a result group.
+type ResultGroup struct {
+	// Group key corresponding to the `GroupBy` fields in the query.
+	Key []field.ValueUnion
```

Comment:
Would `[]field.ValueType` more suffice here? Also, maybe rename to `GroupByKeys`?

Reply:
These are the keys used as the identity of each group so I don't think we can live with just a list of types? `Keys` is probably sufficient given that the struct is already named `ResultGroup`.

Reply:
I think I'm misunderstanding something, ValueUnion looks like:
```
type ValueUnion struct {
	Type         ValueType
	BoolVal      bool
	IntVal       int
	DoubleVal    float64
	StringVal    string
	TimeNanosVal int64
}
```
not sure how we can "key" on it since there's no path info and essentially all we have is `ValueType`. The other variables seem useless for keying unless we want to GroupBy `int=40`? Is that what we're going after?

Reply:
Yeah so the value union is to work around the limitation of lack of generics and no true union support. Imagine you have a query that groups results by a string field `foo` and a int field `bar`, then the keys for each group will be a []ValueUnion whose first element is a value union containing a string value and the second element is a value union containing an int value.

Reply:
TBD, we can either treat them as different groups, or disallow that.